File: elixir/rail-fence-cipher/lib/rail_fence_cipher.ex
defmodule RailFenceCipher do
  @doc """
  Encode a given plaintext to the corresponding rail fence ciphertext
  """
  @spec encode(String.t(), pos_integer) :: String.t()
  def encode(str, 1), do: str
  def encode("", _), do: ""
  def encode(str, rails) do
    str
    |> String.graphemes()
    |> Stream.zip(Stream.concat(0..rails - 1, rails - 2..1) |> Stream.cycle())
    |> Stream.map(fn {value, index} -> {rem(index, rails), value} end)
    |> Enum.reduce(%{}, fn {modulo, value}, acc ->
      Map.update(acc, modulo, [value], fn values -> [value | values] end)
    end)
    |> Map.values()
    |> Stream.map(&Enum.reverse/1)
    |> Stream.flat_map(&(&1))
    |> Enum.join()
  end

  @doc """
  Decode a given rail fence ciphertext to the corresponding plaintext
  """
  @spec decode(String.t(), pos_integer) :: String.t()
  def decode(str, 1), do: str
  def decode("", _), do: ""
  def decode(str, rails) do
    0..rails - 1
    |> Stream.concat(rails - 2..1)
    |> Stream.cycle()
    |> Stream.zip(0..String.length(str) - 1)
    |> Enum.sort_by(&elem(&1, 0))
    |> Stream.zip(String.graphemes(str))
    |> Enum.sort_by(fn {{_, x}, _} -> x end)
    |> Stream.map(&elem(&1, 1))
    |> Enum.join()
  end
end

File: c/rail-fence-cipher/rail_fence_cipher.c
#include "rail_fence_cipher.h"
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>

char *encode(char *text, size_t rails) {
  size_t length = strlen(text);
  char *encoded = malloc(length + 1);
  size_t index = 0;
  for (size_t i = 0; i < rails; ++i) {
    size_t offset1 = 2 * (rails - 1 - i > 0 ? rails - 1 - i : rails - 1);
    size_t offset2 = 2 * i;
    offset1 = offset1 == 0 ? offset2 : offset1;
    offset2 = offset2 == 0 ? offset1 : offset2;
    bool use_offset1 = false;
    for (size_t j = i; j < length; j += (use_offset1 ? offset1 : offset2)) {
      encoded[index++] = text[j];
      use_offset1 = !use_offset1;
    }
  }
  encoded[length] = '\0';
  return encoded;
}

char *decode(char *ciphertext, size_t rails) {
  size_t length = strlen(ciphertext);
  char *decoded = malloc(length + 1);
  size_t index = 0;
  for (size_t i = 0; i < rails; ++i) {
    size_t offset1 = 2 * (rails - 1 - i > 0 ? rails - 1 - i : rails - 1);
    size_t offset2 = 2 * i;
    offset1 = offset1 == 0 ? offset2 : offset1;
    offset2 = offset2 == 0 ? offset1 : offset2;
    bool use_offset1 = false;
    for (size_t j = i; j < length; j += (use_offset1 ? offset1 : offset2)) {
      decoded[j] = ciphertext[index++];
      use_offset1 = !use_offset1;
    }
  }
  decoded[length] = '\0';
  return decoded;
}
